sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -150,8 +161,24 @@ class RocksDB(
           metadata.numKeys
         }
         numKeysOnWritingVersion = numKeys
-        numKeysOnLoadedVersion = numKeys
 
+        for (v <- latestSnapshotVersion + 1 to version) {
+          var changelogReader: ChangelogReader = null
+          try {
+            changelogReader = fileManager.getChangelogReader(v)
+            while (changelogReader.hasNext) {
+              val byteArrayPair = changelogReader.next()
+              if (byteArrayPair.value != null) {
+                put(byteArrayPair.key, byteArrayPair.value)
+              } else {
+                remove(byteArrayPair.key)
+              }
+            }
+          } finally {
+            if (changelogReader != null) changelogReader.close()
+          }
+        }
+        numKeysOnLoadedVersion = numKeysOnWritingVersion
```

Comment:
Is `numKeys` being updated somewhere here ?

Reply:
Yes, during changelog replay, numKeysOnWritingVersion is updated inside put() and remove().